src/FreetypeRenderer.cc

```diff
@@ -256,49 +319,186 @@ std::vector<const Geometry *> FreetypeRenderer::render(const FreetypeRenderer::P
         hb_glyph_position_t *glyph_pos = hb_buffer_get_glyph_positions(hb_buf, &glyph_count);
 
-	GlyphArray glyph_array;
 	for (unsigned int idx = 0; idx < glyph_count; ++idx) {
 		FT_UInt glyph_index = glyph_info[idx].codepoint;
 		error = FT_Load_Glyph(face, glyph_index, FT_LOAD_DEFAULT);
 		if (error) {
-			LOG(message_group::None,Location::NONE,"","Could not load glyph %1$u for char at index %2$u in text '%3$s'",glyph_index,idx,params.text);
+			LOG(message_group::Warning,params.loc,params.documentPath,"Could not load glyph %1$u for char at index %2$u in text '%3$s'",glyph_index,idx,params.text);
 			continue;
 		}
 
 		FT_Glyph glyph;
 		error = FT_Get_Glyph(face->glyph, &glyph);
 		if (error) {
-			LOG(message_group::None,Location::NONE,"","Could not get glyph %1$u for char at index %2$u in text '%3$s'",glyph_index,idx,params.text);
+			LOG(message_group::Warning,params.loc,params.documentPath,"Could not get glyph %1$u for char at index %2$u in text '%3$s'",glyph_index,idx,params.text);
 			continue;
 		}
 		const GlyphData *glyph_data = new GlyphData(glyph, idx, &glyph_pos[idx]);
 		glyph_array.push_back(glyph_data);
 	}
 
-	double width = 0, ascend = 0, descend = 0;
-	for (const auto glyph : glyph_array) {
-		FT_BBox bbox;
-		FT_Glyph_Get_CBox(glyph->get_glyph(), FT_GLYPH_BBOX_GRIDFIT, &bbox);
-		
-		if (HB_DIRECTION_IS_HORIZONTAL(hb_buffer_get_direction(hb_buf))) {
-			double asc = std::max(0.0, bbox.yMax * unscale);
-			double desc = std::max(0.0, -bbox.yMin * unscale);
-			width += glyph->get_x_advance() * params.spacing;
-			ascend = std::max(ascend, asc);
-			descend = std::max(descend, desc);
-		} else {
-			double w_bbox = (bbox.xMax - bbox.xMin) * unscale;
-			width = std::max(width, w_bbox);
-			ascend += glyph->get_y_advance() * params.spacing;
-		}
-	}
-	
-	double x_offset = calc_x_offset(params.halign, width);
-	double y_offset = calc_y_offset(params.valign, ascend, descend);
+    ascent = std::numeric_limits<double>::lowest();
+    descent = std::numeric_limits<double>::max();
+    advance_x = 0;
+    advance_y = 0;
+    left = std::numeric_limits<double>::max();
+    right = std::numeric_limits<double>::lowest();
+    bottom = std::numeric_limits<double>::max();
+    top = std::numeric_limits<double>::lowest();
+
+    bool has_ink = false;
+    for (const auto glyph : glyph_array) {
+        FT_BBox bbox;
+        FT_Glyph_Get_CBox(glyph->get_glyph(), FT_GLYPH_BBOX_GRIDFIT, &bbox);
+
+        // Note that glyphs can extend left of their origin
+        // and right of their advance-width, into the next
+        // glyph's box.  In theory they could extend
+        // arbitrarily through other glyphs' boxes.  Hence the
+        // need for the min/max here.
+
+        // Glyphs with null bounding boxes do not contribute
+        // ink and so do not contribute to the bounding box or
+        // ascent and descent.
+        if (bbox.xMax > bbox.xMin && bbox.yMax > bbox.yMin) {
```

Comment:
Not sure.  Maybe, but not very much.  I think it would turn the following min/max sequences into constructing boxes with the bounding box for the current character, and then using `Eigen::AlignedBox::extend()` to merge them into the current box.  I think it would be very close to the same amount of text.  Whether it would be clearer would depend on whether you think of it as calculating the X and Y positions of the bounding boxes, or as unioning a set of bounding boxes.